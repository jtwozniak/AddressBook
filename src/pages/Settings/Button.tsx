import React from "react"
import styled from "styled-components"
import { darkStyle } from "../../comps/styleHelper"

const StyledButton = styled.button`
  ${darkStyle}
`

type Props = React.PropsWithChildren<{
  onChange: () => void
}>

export const Button = ({ children, onChange }: Props) => (
  <StyledButton type="button" data-testid="apply-button" onClick={onChange}>
    {children}
  </StyledButton>
)
