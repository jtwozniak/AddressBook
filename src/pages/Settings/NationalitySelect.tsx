import React, { ChangeEventHandler } from "react"
import styled from "styled-components"
import { darkStyle } from "../../comps/styleHelper"

const Select = styled.select`
  ${darkStyle}
`

export const OptionsMap = {
  All: "All",
  CH: "Swiss",
  ES: "Spanish",
  FR: "French",
  GB: "United Kingdom",
}

type Props = {
  value: string
  onChange: (n: string) => void
}

export const NationalitySelect = ({ value, onChange }: Props) => (
  <Select
    name="Nationality"
    id="nationality"
    value={value}
    onChange={({
      currentTarget: { value },
    }: React.SyntheticEvent<HTMLSelectElement>) => onChange(value)}
  >
    {Object.entries(OptionsMap).map(([symbol, nationality]) => (
      <option key={symbol} value={symbol}>
        {nationality}
      </option>
    ))}
  </Select>
)
