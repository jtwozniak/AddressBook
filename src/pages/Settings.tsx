import React, { useCallback, useState } from "react"
import styled from "styled-components"
import { useHistory } from "react-router-dom"
import { Button } from "../comps/Button"
import { getNationalityLink, NationalityLink } from "../comps/NationalityLink"
import { NationalitySelect, OptionsMap } from "../comps/NationalitySelect"
import { PageLayout } from "../comps/PageLayout"
import { useNationality } from "../comps/useNationality"

const H2 = styled.h2`
  align-content: center;
  text-align: center;
  margin-top: 5em;
`

export const Settings = () => {
  const initialNationality = useNationality()
  const [nationality, setNationality] = useState(initialNationality)
  const history = useHistory()

  const onApply = useCallback(() => {
    history.push(
      getNationalityLink(
        // this is done to maintain no param in url when All is selected
        OptionsMap.All === nationality ? null : nationality,
        "/settings"
      )
    )
  }, [nationality])

  return (
    <PageLayout>
      <H2>
        Select nationality:
        <NationalitySelect value={nationality} onChange={setNationality} />
        <Button onChange={onApply}>Apply</Button>
      </H2>
      <H2>
        <NationalityLink to={"/"}>Back to main page</NationalityLink>
      </H2>
    </PageLayout>
  )
}
