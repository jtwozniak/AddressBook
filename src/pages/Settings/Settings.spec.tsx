import React from "react"
import { act, fireEvent } from "@testing-library/react"
import { renderWithRouter } from "../../testUtils/renderWithRouter"
import { Settings } from "./Settings"

// we don't have main router setup so useParams will not extract url params
// so we mock useParams accordingly
jest.mock("../../comps/useNationality", () => ({
  useNationality: () => ({
    nationality: "GB",
  }),
}))

describe("Settings", () => {
  it("Page snapshot", () => {
    const { container, history } = renderWithRouter(<Settings />)
    expect(container).toMatchSnapshot()
    expect(history.location.pathname).toStrictEqual("/")
  })

  it("Nationality settings are applied in pathname", () => {
    const { history, getByTestId } = renderWithRouter(<Settings />, {
      route: "/settings",
    })

    act(() => {
      fireEvent.change(getByTestId("nationality-select"), {
        target: { value: "CH" },
      })
    })

    expect(history.location.pathname).toStrictEqual("/settings")

    act(() => {
      getByTestId("apply-button").click()
    })
    expect(history.location.pathname).toStrictEqual("/CH/settings")
  })

  it("Nationality settings are used in home path", () => {
    const { history, getByTestId, getByText } = renderWithRouter(<Settings />, {
      route: "/settings",
    })

    act(() => {
      fireEvent.change(getByTestId("nationality-select"), {
        target: { value: "GB" },
      })
    })

    act(() => {
      getByTestId("apply-button").click()
    })

    act(() => {
      getByText("Back to main page").click()
    })

    expect(history.location.pathname).toStrictEqual("/GB/")
  })
})
