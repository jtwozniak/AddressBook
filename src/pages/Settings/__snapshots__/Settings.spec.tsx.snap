// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Settings Page snapshot 1`] = `
.c5 {
  background-color: black;
  color: white;
  margin-left: 1em;
}

.c4 {
  background-color: black;
  color: white;
  margin-left: 1em;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  height: 100vh;
  width: 100%;
}

.c2 {
  width: 100%;
  height: 100%;
  padding: 2em;
}

.c3 {
  -webkit-align-content: center;
  -ms-flex-line-pack: center;
  align-content: center;
  text-align: center;
  margin-top: 5em;
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1"
    >
      <h1>
        Address Book
      </h1>
      <a
        data-testid="link"
        href="/GB/settings"
      >
        <h1>
          Settings
        </h1>
      </a>
    </div>
    <main
      class="c2"
    >
      <h2
        class="c3"
      >
        Select nationality:
        <select
          class="c4"
          data-testid="nationality-select"
          id="nationality-select"
          name="Nationality"
        >
          <option
            data-testid="nationality-option"
            value="All"
          >
            All
          </option>
          <option
            data-testid="nationality-option"
            value="CH"
          >
            Swiss
          </option>
          <option
            data-testid="nationality-option"
            value="ES"
          >
            Spanish
          </option>
          <option
            data-testid="nationality-option"
            value="FR"
          >
            French
          </option>
          <option
            data-testid="nationality-option"
            value="GB"
          >
            United Kingdom
          </option>
        </select>
        <button
          class="c5"
          data-testid="apply-button"
          type="button"
        >
          Apply
        </button>
      </h2>
      <h2
        class="c3"
      >
        <a
          data-testid="link"
          href="/GB/"
        >
          Back to main page
        </a>
      </h2>
    </main>
  </div>
</div>
`;
